addons/binding/org.openhab.binding.connectsdk/src/main/java/org/openhab/binding/connectsdk/internal/discovery/ConnectSDKDiscovery.java

```diff
@@ -0,0 +1,206 @@
+        super(ConnectSDKBindingConstants.SUPPORTED_THING_TYPES_UIDS, 60, true);
+        DiscoveryManager.init(this);
+    }
+
+    @Override
+    protected void activate(Map<String, Object> configProperties) {
+        logger.info(configProperties.toString());
+        localInetAddresses = findLocalInetAddresses((String) configProperties.get("localIP"));
+        Util.init(AbstractDiscoveryService.scheduler);
+        discoveryManager = DiscoveryManager.getInstance();
+        discoveryManager.setPairingLevel(DiscoveryManager.PairingLevel.ON);
+        discoveryManager.addListener(this);
+        super.activate(configProperties); // starts background discovery
+    }
+
+    @Override
+    protected void deactivate() {
+        super.deactivate(); // stops background discovery
+        discoveryManager.removeListener(this);
+        discoveryManager = null;
+        DiscoveryManager.destroy();
+        Util.uninit();
+    }
+
+    // @Override
+    @Override
+    protected void startScan() {
+        // no adhoc scanning. Discovery Service runs in background, but re-discover all known devices incase they were
+        // deleted from the inbox.
+        for (ConnectableDevice d : discoveryManager.getCompatibleDevices().values()) {
+            thingDiscovered(createDiscoveryResult(d));
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        discoveryManager.start();
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        discoveryManager.stop();
+    }
+
+    // DiscoveryManagerListener
+
+    @Override
+    public void onDeviceAdded(DiscoveryManager manager, ConnectableDevice device) {
+        thingDiscovered(createDiscoveryResult(device));
+    }
+
+    @Override
+    public void onDeviceUpdated(DiscoveryManager manager, ConnectableDevice device) {
+        logger.info("Device updated: {}", device);
+        thingRemoved(createThingUID(device));
+        thingDiscovered(createDiscoveryResult(device));
+    }
+
+    @Override
+    public void onDeviceRemoved(DiscoveryManager manager, ConnectableDevice device) {
+        logger.info("Device removed: {}", device);
+        thingRemoved(createThingUID(device));
+    }
+
+    @Override
+    public void onDiscoveryFailed(DiscoveryManager manager, ServiceCommandError error) {
+        logger.warn("Discovery Failed {}", error.getMessage());
+
+    }
+
+    // Helpers for DiscoveryManagerListener Impl
+    private DiscoveryResult createDiscoveryResult(ConnectableDevice device) {
+        ThingUID thingUID = createThingUID(device);
+        DiscoveryResult result = DiscoveryResultBuilder.create(thingUID)
+                .withProperty(PROPERTY_IP_ADDRESS, device.getIpAddress()).withLabel(device.getFriendlyName()).build();
+        return result;
+    }
+
+    private ThingUID createThingUID(ConnectableDevice device) {
+        return new ThingUID(THING_TYPE_WebOSTV, device.getIpAddress().replace('.', '_'));
```

Comment:
Is there any other parameter which is not changeable, like serial number, that can be used for building thing ID? This should allow for better identification of a device even with changing IP, or device could use DHPC freely. There seems to be an ID in ConnectableDevice, but not sure if this is it: 'universally unique id of this particular ConnectableDevice object, persists between sessions in ConnectableDeviceStore for connected devices'

Reply:
ConnectSDK uses IPs as IDs. I am just sticking with the same pattern.

Reply:
Connect SDK device registry is based on IPs as primary identifier. I therefore followed along the same concept.

Reply:
OK